### PR TITLE
idtools: add support for libsubid

### PR DESCRIPTION
**Pull request by giuseppe**
when building with cgo, add support for libsubid to read the
additional sub IDs for the user instead of parsing the /etc/sub?id
files.

Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Thanks @giuseppe When do you think libsubuid will show up?

**Comment by giuseppe** [comment]
> Thanks @giuseppe When do you think libsubuid will show up?

I am not sure about that, AFAIK there wasn't even a release upstream yet.

@alexey-tikhonov do you know when a new shadow-utils can land in Fedora?

**Comment by alexey-tikhonov** [comment]
It is actually in Fedora 35 already: https://bodhi.fedoraproject.org/updates/FEDORA-2021-8483b52301

(but most probably will need an update when https://github.com/shadow-maint/shadow/issues/325 is fixed)

**Comment by giuseppe** [comment]
thanks!  Sorry I've missed it before.

I've tested the current patch with the package available for F35 and it works fine

**Comment by rhatdan** [comment]
What do we need to do to get this in and able to work on distros without libsubid?

**Comment by giuseppe** [comment]
> What do we need to do to get this in and able to work on distros without libsubid?

Probably using a build tag would be the easiest and we can use it in Rawhide

**Comment by alexey-tikhonov** [comment]
Btw, there were several changes in libsubid (including https://github.com/shadow-maint/shadow/commit/3d670ba7ed58f910f1395569ea14cd39fd046cc7)
We will rebuild package in Rawhide soon (probably after shadow-utils upstream release)

**Comment by rhatdan** [comment]
Can you add something like:
```
cat hack/libsubid_tag.sh
#!/usr/bin/env bash
if test $(${GO:-go} env GOOS) != "linux" ; then
	exit 0
fi
tmpdir="$PWD/tmp.$RANDOM"
mkdir -p "$tmpdir"
trap 'rm -fr "$tmpdir"' EXIT
cc -o "$tmpdir"/libdm_tag -ldevmapper -x c - > /dev/null 2> /dev/null << EOF
#include <"shadow/subid.h">
int main() {
        subid_range *ranges;
	err := get_subuid_ranges("root", &ranges);
	return 0;
}
EOF
if test $? -eq 0 ; then
	echo libsubid
fi
```

**Comment by rhatdan** [comment]
Also modify the Makefile with

AUTOTAGS := $(shell ./hack/btrfs_tag.sh) $(shell ./hack/libdm_tag.sh) $(shell ./hack/libsubid_tag.sh)

**Comment by rhatdan** [comment]
Perhaps we should make the Tag a negative as well so it will always get compiled in unless users specify don't use it.

**Comment by giuseppe** [comment]
switched the logic

File: hack/libsubid_tag.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+if test $(${GO:-go} env GOOS) != "linux" ; then
+	exit 0
```
Comment (rhatdan) [inline hack/libsubid_tag.sh:3]: This should also do 
 	echo no_libsubid

**Comment by rhatdan** [comment]
One issue, and then LGTM
Is this still a WIP, or something we can merge now?

File: hack/libsubid_tag.sh (hunk 1 above)
Comment (giuseppe) [reply hack/libsubid_tag.sh:4]: amended

**Comment by giuseppe** [comment]
>  Is this still a WIP, or something we can merge now?

it is a WIP because the libsubid API were not released yet.

@alexey-tikhonov is there any risk the API will change again or can we merge this PR?

**Comment by rhatdan** [comment]
LGTM
@saschagrunert @vrothberg @nalind @TomSweeneyRedHat PTAL

**Comment by rhatdan** [comment]
It would be nice to get a version of Podman into Rawhide that used libsubid, to make sure all of the plumbing works.

**Comment by rhatdan** [comment]
@alexey-tikhonov PTAL

File: pkg/idtools/idtools_supported.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+{
+    return ranges[i];
+}
+*/
+import "C"
+
+func readSubid(username string, isUser bool) (ranges, error) {
+	var ret ranges
+	if username == "ALL" {
+		return nil, fmt.Errorf("username ALL not supported")
```
Comment (saschagrunert) [inline pkg/idtools/idtools_supported.go:25]:
```suggestion
		return nil, errors.New("username ALL not supported")
```

**Review by saschagrunert — APPROVED** [review]
Just a non-blocking nit, otherwise LGTM

File: pkg/idtools/idtools_supported.go (hunk 2 above)
Comment (giuseppe) [reply pkg/idtools/idtools_supported.go]: thanks, fixed now

**Comment by alexey-tikhonov** [comment]
> it is a WIP because the libsubid API were not released yet.
> 
> @alexey-tikhonov is there any risk the API will change again or can we merge this PR?

To the best of my knowledge, I don't anticipate `libsubid` API changes in a near future. (And this is already merged in C9S: https://gitlab.com/redhat/centos-stream/rpms/shadow-utils/-/commit/fbf9d3a3ea5ea94a96eb747034c5ca9a25095633)

But that's true that there is no upstream release yet and I'm not authoritative source here.
@hallyn, what's your current plan with regards to upstream release of shadow-utils?



> @alexey-tikhonov PTAL

Looks good to me.

**Comment by alexey-tikhonov** [comment]
JFYI: https://github.com/shadow-maint/shadow/releases/tag/v4.9

**Comment by ikerexxe** [comment]
I'm working on rebasing Fedora rawhide to the new version of shadow-utils. Once it is ready I'll let you know.

**Comment by rhatdan** [comment]
Awesome news, Are you planning on releasing this to Fedora 34?  We will need this released/backported to RHEL8.6, also.

**Comment by alexey-tikhonov** [comment]
> Awesome news, Are you planning on releasing this to Fedora 34? We will need this released/backported to RHEL8.6, also.

RHEL8.6 - sure, Fedora 34 - not planned at the moment.

**Comment by rhatdan** [comment]
@giuseppe Now that libsubuid is in Rawhide, can we move this out of draft?

**Comment by giuseppe** [comment]
rebased

**Comment by ikerexxe** [comment]
Rawhide update is ready https://bodhi.fedoraproject.org/updates/FEDORA-2021-3466bdabc1

**Pull request merged**